lib/model/channel.dart

```diff
@@ -572,6 +641,65 @@ class ChannelStoreImpl extends HasUserStore with ChannelStore {
       forStream[event.topicName] = visibilityPolicy;
     }
   }
+
+  /// Handle a [MessageEvent], returning whether listeners should be notified.
+  bool handleMessageEvent(MessageEvent event) {
+    if (event.message is! StreamMessage) return false;
+    final StreamMessage(streamId: channelId, :topic) = event.message as StreamMessage;
+
+    final latestMessageIdsByTopic = _latestMessageIdsByChannelTopic[channelId];
+    if (latestMessageIdsByTopic == null) {
+      // We're not tracking this channel's topics yet.
+      // We start doing that when [fetchTopics] is called,
+      // and we fill in all the topics at that time.
+      return false;
+    }
+
+    final currentLatestMessageId = latestMessageIdsByTopic[topic];
+    if (currentLatestMessageId != null && currentLatestMessageId >= event.message.id) {
+      // The event raced with a message fetch.
+      return false;
+    }
+    latestMessageIdsByTopic[topic] = event.message.id;
+    return true;
+  }
+
+  /// Handle an [UpdateMessageEvent], returning whether listeners should be
+  /// notified.
+  bool handleUpdateMessageEvent(UpdateMessageEvent event) {
+    if (event.moveData == null) return false;
+    final UpdateMessageMoveData(
+      :origStreamId, :origTopic, :newStreamId, :newTopic, :propagateMode,
+    ) = event.moveData!;
+    bool shouldNotify = false;
+
+    final origLatestMessageIdsByTopics = _latestMessageIdsByChannelTopic[origStreamId];
+    switch (propagateMode) {
+      case PropagateMode.changeOne:
+      case PropagateMode.changeLater:
+        // We can't know the new `maxId` for the original topic.
+        // Shrug; leave it unchanged. (See dartdoc of [getChannelTopics],
+        // where we call out this possibility that `maxId` is incorrect.
+        break;
+      case PropagateMode.changeAll:
+        if (origLatestMessageIdsByTopics != null) {
+          origLatestMessageIdsByTopics.remove(origTopic);
+          shouldNotify = true;
+        }
+    }
+
+    final newLatestMessageIdsByTopics = _latestMessageIdsByChannelTopic[newStreamId];
+    if (newLatestMessageIdsByTopics != null) {
+      final movedMaxId = event.messageIds.max;
```

Comment:
```suggestion
      // TODO(server-11) rely on `event.messageIds` being sorted, to avoid this linear scan
      final movedMaxId = event.messageIds.max;
```